README.md
# keep-alive

 This module allows you to set per-connection keep alive parameters on windows and linux enviroments.
 For more information on keep alive signals see <https://en.wikipedia.org/wiki/Keepalive>.
 See also <https://tldp.org/HOWTO/html_single/TCP-Keepalive-HOWTO/> for a linux specific implementation.

 The module is meant to be used in conjuction with the "network" package. Speciffically, all functions require
 a socket file descriptor argument which can be obtained by the withFdSocket function:

```hs
 -- sock is a Socket type
 withFdSocket sock $ \fd do
    before <- getKeepAliveOnOff fd
    print before -- False
    -- set keep alive on, idle 60 seconds, interval 2 seconds
    rlt <- setKeepAlive fd $ KeepAlive True 60 2
    case rlt of
        Left err -> print err
        Right () -> return ()
    after <- getKeepAliveOnOff fd
    print after -- True
```


ChangeLog.md
# Changelog for keep-alive

## 0.1.1.0

- Fix build errors

## 0.2.0.0

- add return code to OTHER_KEEPALIVE_ERROR

## Unreleased changes


src/Lib.hs
{-# LANGUAGE CPP,  ForeignFunctionInterface #-}
module Lib 
    ( KeepAlive (..)
    , setKeepAlive
    , getKeepAliveOnOff
    ) where

import           Foreign
import           Foreign.C
import           LibForeign
import Data.Word (Word32)


data KeepAlive = KeepAlive
    { keepOnOff :: Bool
    , keepIdle  :: Word32
    , keepIntvl :: Word32
    }
    deriving (Show, Eq, Ord)

data KeepAliveError
    = WSA_IO_PENDING
    | WSA_OPERATION_ABORTED
    | WSAEFAULT
    | WSAEINPROGRESS
    | WSAEINTR
    | WSAEINVAL
    | WSAENETDOWN
    | WSAENOPROTOOPT
    | WSAENOTSOCK
    | WSAEOPNOTSUPP
    | EBADF
      -- ^ The socket argument is not a valid file descriptor.
    | EDOM
      -- ^ The send and receive timeout values are too big to
      -- fit into the timeout fields in the socket structure.
    | EINVAL
      -- ^ The specified option is invalid at the specified socket level
      -- or the socket has been shut down.
    | EISCONN
      -- ^ The socket is already connected, and a specified option cannot
      -- be set while the socket is connected.
    | ENOPROTOOPT
      -- ^ The option is not supported by the protocol.
    | ENOTSOCK
      -- ^ The socket argument does not refer to a socket.
    | ENOMEM
      -- ^ There was insufficient memory available for the operation to complete.
    | ENOBUFS
      -- ^ Insufficient resources are available in the system to complete the call.
    | OTHERKEEPALIVEERROR
    deriving (Show, Eq, Ord)

setKeepAlive :: CInt -> KeepAlive -> IO ( Either KeepAliveError ())
setKeepAlive fd (KeepAlive onoff idle intvl) = do
    rlt <- setKeepAlive_ fd (cFromBool onoff) idle intvl
    return $ case rlt of
        0 -> Right ()
        997 -> Left WSA_IO_PENDING
        995 -> Left WSA_OPERATION_ABORTED
        10014 -> Left WSAEFAULT
        10036 -> Left WSAEINPROGRESS
        10004 -> Left WSAEINTR
        10022 -> Left WSAEINVAL
        10050 -> Left WSAENETDOWN
        10042 -> Left WSAENOPROTOOPT
        10038 -> Left WSAENOTSOCK
        10045 -> Left WSAEOPNOTSUPP
        9 -> Left EBADF
        33 -> Left EDOM
        22 -> Left EINVAL
        106 -> Left EISCONN
        92 -> Left ENOPROTOOPT
        88 -> Left ENOTSOCK
        12 -> Left ENOMEM
        105 -> Left ENOBUFS
        _ -> Left OTHERKEEPALIVEERROR

getKeepAliveOnOff :: CInt -> IO Bool
getKeepAliveOnOff fd = 
    cToBool . fromInteger . toInteger <$> getKeepAliveOnOff_ fd
 
cToBool :: Int -> Bool
cToBool x
    | x == 0 = False
    | otherwise = True

cFromBool :: Bool -> Word32
cFromBool True = 1
cFromBool False = 0



app/Main.hs
module Main where

import Network.Socket.KeepAlive
import qualified Network.Socket             as S
import           Data.IP.Internal           (IPv4)
import Data.IP (toHostAddress)
import Control.Exception (bracket)
import Control.Concurrent (threadDelay)
import Foreign (sizeOf, void, with)

main :: IO ()
main = ffi

ffi :: IO()
ffi = do
    let ip = read "192.168.1.35"
    let port = 5502
    let addr = getAddr ip port
    withSocket addr $ \s -> do
        S.withFdSocket s $ \fd -> do
            before <- getKeepAliveOnOff fd
            print before
            rlt <- setKeepAlive fd test
            case rlt of
                Left err -> print err
                Right () -> return ()
            after <- getKeepAliveOnOff fd
            print after
        threadDelay 10000000000    
        return ()

test :: KeepAlive
test = KeepAlive True 10 15

withSocket :: S.SockAddr -> (S.Socket -> IO a) -> IO a
withSocket addr = bracket (connectTCP addr) close
  where close s = S.gracefulClose s 1000

connectTCP :: S.SockAddr -> IO S.Socket
connectTCP addr = do
    putStrLn ("Connecting to " ++ show addr ++ "...")
    s <- S.socket S.AF_INET S.Stream S.defaultProtocol
    S.connect s addr
    putStrLn "connected"
    return s

getAddr :: IPv4 -> Int -> S.SockAddr
getAddr ip portNum = S.SockAddrInet (fromIntegral portNum) (toHostAddress ip)

src/Network/Socket/KeepAlive.hs
{-# LANGUAGE CPP #-}

-----------------------------------------------------------------------------
-- |
-- This module allows you to set per-connection keep alive parameters on windows and linux enviroments.
-- For more information on keep alive signals see https://en.wikipedia.org/wiki/Keepalive.
-- See also https://tldp.org/HOWTO/html_single/TCP-Keepalive-HOWTO/ for a linux specific implementation.
--
-- The module is meant to be used in conjuction with the "network" package. However, in order to ensure adaptability, all functions require
-- a socket file descriptor instead of an implementation dependent socket type. For the network package, such a  descriptor can be obtained
-- with the withFdSocket function:
--
-- > -- sock is a Socket type
-- > withFdSocket sock $ \fd -> do
-- >    before <- getKeepAliveOnOff fd
-- >    print before -- False
-- >    -- set keep alive on, idle 60 seconds, interval 2 seconds
-- >    rlt <- setKeepAlive fd $ KeepAlive True 60 2
-- >    case rlt of
-- >        Left err -> print err
-- >        Right () -> return ()
-- >    after <- getKeepAliveOnOff fd
-- >    print after -- True
--
-- Please note that only the envocing process can manipulate sockets based on their file descriptors.

module Network.Socket.KeepAlive
    ( KeepAlive (..)
    , KeepAliveError (..)
    , setKeepAlive
    , getKeepAliveOnOff
    ) where

import           Foreign    (Word32)
import           Foreign.C  (CInt)
import           LibForeign (getKeepAliveOnOff_, setKeepAlive_)


-- | The main data structure defining keep alive parameters
data KeepAlive = KeepAlive
    { kaOnOff :: Bool
    -- ^ Turns on / off keep alive probes
    , kaIdle  :: Word32
    -- ^ The interval in seconds between the last data packet sent and the first keep alive probe
    , kaIntvl :: Word32
    -- ^ The interval in seconds between subsequential keepalive probes
    }
    deriving (Show, Eq, Ord)

-- | Errors starting with WSA are windows specific
data KeepAliveError
    = WSA_IO_PENDING
    | WSA_OPERATION_ABORTED
    | WSAEFAULT
    | WSAEINPROGRESS
    | WSAEINTR
    | WSAEINVAL
    | WSAENETDOWN
    | WSAENOPROTOOPT
    | WSAENOTSOCK
    | WSAEOPNOTSUPP
    | EBADF
    | EDOM
    | EINVAL
    | EISCONN
    | ENOPROTOOPT
    | ENOTSOCK
    | ENOMEM
    | ENOBUFS
    | OTHER_KEEPALIVE_ERROR CInt
    deriving (Show, Eq, Ord)

-- | Set keep alive parameters for the current socket
setKeepAlive ::
    CInt
    -- ^ Socket file descriptor
    -> KeepAlive
    -- ^ Keep alive parameters
    -> IO ( Either KeepAliveError ())
setKeepAlive fd (KeepAlive onoff idle intvl) = do
    rlt <- setKeepAlive_ fd (cFromBool onoff) idle intvl
    return $ case rlt of
        0     -> Right ()
        997   -> Left WSA_IO_PENDING
        995   -> Left WSA_OPERATION_ABORTED
        10014 -> Left WSAEFAULT
        10036 -> Left WSAEINPROGRESS
        10004 -> Left WSAEINTR
        10022 -> Left WSAEINVAL
        10050 -> Left WSAENETDOWN
        10042 -> Left WSAENOPROTOOPT
        10038 -> Left WSAENOTSOCK
        10045 -> Left WSAEOPNOTSUPP
        9     -> Left EBADF
        33    -> Left EDOM
        22    -> Left EINVAL
        106   -> Left EISCONN
        92    -> Left ENOPROTOOPT
        88    -> Left ENOTSOCK
        12    -> Left ENOMEM
        105   -> Left ENOBUFS
        other -> Left $ OTHER_KEEPALIVE_ERROR other

-- | Returns True if keep alive is active for the specified socket
getKeepAliveOnOff ::
    CInt
    -- ^ Socket file descriptor
    -> IO Bool
getKeepAliveOnOff fd =
    cToBool <$> getKeepAliveOnOff_ fd

cToBool :: CInt -> Bool
cToBool x
    | x == 0 = False
    | otherwise = True

cFromBool :: Bool -> Word32
cFromBool True  = 1
cFromBool False = 0

